include/maths/CBoostedTreeFactory.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_maths_CBoostedTreeFactory_h
+#define INCLUDED_ml_maths_CBoostedTreeFactory_h
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+#include <core/CPackedBitVector.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTree.h>
+
+#include <memory>
+#include <numeric>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace ml {
+namespace maths {
+
+class CNode;
+class CBoostedTree;
+
+//! SimpleFactory for CBoostedTree object
+class CBoostedTreeFactory final {
+public:
+    using TBoostedTreeUPtr = std::unique_ptr<CBoostedTree>;
+
+public:
+    //! Construct a boosted tree object from parameters
+    static CBoostedTreeFactory constructFromParameters(std::size_t numberThreads,
+                                                       std::size_t dependentVariable,
+                                                       CBoostedTree::TLossFunctionUPtr loss);
+
+    //! Set the number of folds to use for estimating the generalisation error.
+    CBoostedTreeFactory& numberFolds(std::size_t folds);
+    //! Set the lambda regularisation parameter.
+    CBoostedTreeFactory& lambda(double lambda);
+    //! Set the gamma regularisation parameter.
+    CBoostedTreeFactory& gamma(double gamma);
+    //! Set the amount we'll shrink the weights on each each iteration.
+    CBoostedTreeFactory& eta(double eta);
+    //! Set the maximum number of trees in the ensemble.
+    CBoostedTreeFactory& maximumNumberTrees(std::size_t maximumNumberTrees);
+    //! Set the fraction of features we'll use in the bag to build a tree.
+    CBoostedTreeFactory& featureBagFraction(double featureBagFraction);
+    //! Set the maximum number of optimisation rounds we'll use for hyperparameter
+    //! optimisation per parameter.
+    CBoostedTreeFactory& maximumOptimisationRoundsPerHyperparameter(std::size_t rounds);
+    //! Set the callback function for progress monitoring.
+    CBoostedTreeFactory& progressCallback(CBoostedTree::TProgressCallback callback);
+    //! Set the reference to the data frame.
+    CBoostedTreeFactory& frame(core::CDataFrame& frame);
+    //! Returns the reference to the CBoostedTree object before this was completely
+    //! initialized. This is useful since some functions, e.g estimateMemoryUsage has to be called before
+    //! data frame reference was passed
+    const CBoostedTree& incompleteTreeObject() const;
```

Comment:
I understand why you want to do this, but I don't think it is a good choice in this case. The only thing this object can do is return the memory usage, so it is like wrapping up that calculation in an object whose interface can't be supported. It presumably also means that we have to be able to create a version of `CBoostedTree` with nullable type for `frame` and then have runtime checks that this exists when for example `train` is called. Let's see if we can come up with a better solution offline.

Reply:
Changed in 366750d